Use spartakus to report usage metrics

jlewi opened the pull request:
* Use Kubernetes reporting tool (spartakus) to report anonymous statistics about Kubeflow usage such as basic cluster stats.

* This is optional and we're making it opt in for now.

* One current limitation is that there's no easy way to give each kubeflow deployment a unique, random id, so it will be hard to distinguish different deployments.
    * users can manually assign a unique id.
    * We could potentially modify spartakus (or the Docker image) to generate a random id. The one downside of this is that the id would be regenerated if the pod restarts.

* Related to #55


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/kubeflow/175)
<!-- Reviewable:end -->

ankushagarwal commented:
Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[a discussion](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5vA25B-YqduVlm7k95:-L5vA25B-YqduVlm7k96:b-hzdh3a) (no related file):*
nit: Is it possible to use double quotes everywhere instead of a mix of single quotes and double quotes
[This jsonnet issue](https://github.com/google/jsonnet/issues/72) seems to suggest it is better to use single quotes in jsonnet

---

*[README.md, line 93 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5v-_Tw8o6-gtsXx7gO:-L5v-_Tw8o6-gtsXx7gP:bdyz09q) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/README.md#L93)):*
> ```Markdown
> For more detailed instructions about how to use Kubeflow, please refer to the [user guide](user_guide.md).
> 
> +**Important** By default Kubeflow will collect **anonymous** user data to help us improve Kubeflow; for more information including instructions for disabling it please refer to the [user_guide](user_guide#Usage Reporting)
> ```

Remove `+` at the beginning of the sentence

---

*[user_guide.md, line 82 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5v1CXuELMZXzABYjpl:-L5v1CXuELMZXzABYjpm:b-co1n98) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/user_guide.md#L82)):*
> ```Markdown
> ### Usage Reporting
> 
> By default, Kubeflow will report **anonymous** usage data using [spartakus](https://github.com/kubernetes-incubator/spartakus), Kubernetes' reporting tool. Spartakus **does not report any presonal information** see [here](https://github.com/kubernetes-incubator/spartakus) for more detail.
> ```

nit: `presonal` typo

---

*[user_guide.md, line 82 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5v1GOq2hTI0TaB98bc:-L5v1GOq2hTI0TaB98bd:b-afuri2) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/user_guide.md#L82)):*
> ```Markdown
> ### Usage Reporting
> 
> By default, Kubeflow will report **anonymous** usage data using [spartakus](https://github.com/kubernetes-incubator/spartakus), Kubernetes' reporting tool. Spartakus **does not report any presonal information** see [here](https://github.com/kubernetes-incubator/spartakus) for more detail.
> ```

nit: replace `information see here...` with `information. See here...`

---

*[kubeflow/core/all.libsonnet, line 24 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5v6ji83PWWuKtEAHyd:-L5v6ji83PWWuKtEAHye:b-rhpdg6) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/kubeflow/core/all.libsonnet#L24)):*
> //local tfJobImage = import 'param://tfJobImage';
>     // local tfDefaultImage = import 'param://tfDefaultImage';
>     // local tfJobUiServiceType = import 'param://tfJobUiServiceType';
>     // local jupyterHubServiceType = import 'param://jupyterHubServiceType';

These should be removed

---

*[kubeflow/core/all.libsonnet, line 42 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5vBo0_2sawiRRpPh74:-L5vBo0_2sawiRRpPh75:bo8uqow) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/kubeflow/core/all.libsonnet#L42)):*
> // Create a list of the resources needed for a particular disk
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>
>     local diskToList = function(diskName) [
>       nfs.parts(namespace, name,).diskResources(diskName).storageClass,
>       nfs.parts(namespace, name,).diskResources(diskName).volumeClaim,
>       nfs.parts(namespace, name,).diskResources(diskName).service,
>       nfs.parts(namespace, name,).diskResources(diskName).provisioner,
>     ],
</details>
>  
>     local allDisks = std.flattenArrays(std.map(diskToList, diskNames)),
>  
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>
>     local nfsComponents =
>       if std.length(allDisks) > 0 then
>         [
>           nfs.parts(namespace, name).serviceAccount,
>           nfs.parts(namespace, name).role,
>           nfs.parts(namespace, name).roleBinding,
>           nfs.parts(namespace, name).clusterRoleBinding,
>         ] + allDisks
>       else
>         [],
</details>

Can this logic be moved to nfs.libsonnet

---

*[kubeflow/core/test.jsonnet, line 1 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5v35Y29vVI9EpqygVS:-L5v35Y29vVI9EpqygVT:biswynz) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/kubeflow/core/test.jsonnet#L1)):*
> ```
> local util = import 'util.libsonnet';
> ```

This file is no longer needed after https://github.com/kubeflow/kubeflow/commit/2c39c9d94ebaedbd3470f1fcebad3c11fcfe37bf

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175)*
<!-- Sent from Reviewable.io -->

ankushagarwal commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[a discussion](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5yYgoh8bJLRKJSUcbR:-L5yYgoh8bJLRKJSUcbS:byivxzj) (no related file):*
Have you considered putting the spartakus deployment under the kube-system namespace?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175)*
<!-- Sent from Reviewable.io -->

jlewi commented:
Review status: 7 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

*[a discussion](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5vA25B-YqduVlm7k95:-L5yvpmN5IGMZYbPEf6a:b-bhvyy) (no related file):*
<details><summary><i>Previously, ankushagarwal (Ankush Agarwal) wrote…</i></summary><blockquote>

nit: Is it possible to use double quotes everywhere instead of a mix of single quotes and double quotes
[This jsonnet issue](https://github.com/google/jsonnet/issues/72) seems to suggest it is better to use single quotes in jsonnet
</blockquote></details>

Updated a bunch of places to double quotes. Its not clear to me from the linked issue why double might be better.

If we're serious about wanting to enforce consistency we should look into using a lint check. 

`jsonnet fmt` can be used to autoformat and provide info about some formatting errors; don't think it handles quotes though.

---

*[a discussion](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5yYgoh8bJLRKJSUcbR:-L5ywL3L7VDftogFEvpk:bngg15l) (no related file):*
<details><summary><i>Previously, ankushagarwal (Ankush Agarwal) wrote…</i></summary><blockquote>

Have you considered putting the spartakus deployment under the kube-system namespace?
</blockquote></details>

I think we should treat is as a component of your kubeflow deployment, so I think it belongs in your kubeflow namespace.

For example if we tear down the namespace containing kubeflow we should stop sending metrics about your deployment.

---

*[README.md, line 93 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5v-_Tw8o6-gtsXx7gO:-L5ywkta3BVXz7W5ivKV:b-896fix) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/README.md#L93)):*
<details><summary><i>Previously, ankushagarwal (Ankush Agarwal) wrote…</i></summary><blockquote>

Remove `+` at the beginning of the sentence
</blockquote></details>

Done.

---

*[user_guide.md, line 82 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5v1CXuELMZXzABYjpl:-L5yws0yCM3gxe56uL5m:b-896fix) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/user_guide.md#L82)):*
<details><summary><i>Previously, ankushagarwal (Ankush Agarwal) wrote…</i></summary><blockquote>

nit: `presonal` typo
</blockquote></details>

Done.

---

*[user_guide.md, line 82 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5v1GOq2hTI0TaB98bc:-L5yx1qH5XUkgYZu-0p7:b-896fix) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/user_guide.md#L82)):*
<details><summary><i>Previously, ankushagarwal (Ankush Agarwal) wrote…</i></summary><blockquote>

nit: replace `information see here...` with `information. See here...`
</blockquote></details>

Done.

---

*[kubeflow/core/all.libsonnet, line 24 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5v6ji83PWWuKtEAHyd:-L5yx3Z54xdnx7W7uRtX:b-896fix) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/kubeflow/core/all.libsonnet#L24)):*
<details><summary><i>Previously, ankushagarwal (Ankush Agarwal) wrote…</i></summary><blockquote>

> //local tfJobImage = import 'param://tfJobImage';
>     // local tfDefaultImage = import 'param://tfDefaultImage';
>     // local tfJobUiServiceType = import 'param://tfJobUiServiceType';
>     // local jupyterHubServiceType = import 'param://jupyterHubServiceType';

These should be removed
</blockquote></details>

Done.

---

*[kubeflow/core/all.libsonnet, line 42 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5vBo0_2sawiRRpPh74:-L5z3JRd3iYBxSbn0nTn:b-896fix) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/kubeflow/core/all.libsonnet#L42)):*
<details><summary><i>Previously, ankushagarwal (Ankush Agarwal) wrote…</i></summary><blockquote>

> // Create a list of the resources needed for a particular disk
>     local diskToList = function(diskName) [
>       nfs.parts(namespace, name,).diskResources(diskName).storageClass,
>       nfs.parts(namespace, name,).diskResources(diskName).volumeClaim,
>       nfs.parts(namespace, name,).diskResources(diskName).service,
>       nfs.parts(namespace, name,).diskResources(diskName).provisioner,
>     ],
>  
>     local allDisks = std.flattenArrays(std.map(diskToList, diskNames)),
>  
>     local nfsComponents =
>       if std.length(allDisks) > 0 then
>         [
>           nfs.parts(namespace, name).serviceAccount,
>           nfs.parts(namespace, name).role,
>           nfs.parts(namespace, name).roleBinding,
>           nfs.parts(namespace, name).clusterRoleBinding,
>         ] + allDisks
>       else
>         [],

Can this logic be moved to nfs.libsonnet
</blockquote></details>

Done.

---

*[kubeflow/core/test.jsonnet, line 1 at r1](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175#-L5v35Y29vVI9EpqygVS:-L5yxg1F53pbrnMD5BSM:b-70tigt) ([raw file](https://github.com/kubeflow/kubeflow/blob/7b470698eb12c4cfe47ee96cafc3836af310b78e/kubeflow/core/test.jsonnet#L1)):*
<details><summary><i>Previously, ankushagarwal (Ankush Agarwal) wrote…</i></summary><blockquote>

This file is no longer needed after https://github.com/kubeflow/kubeflow/commit/2c39c9d94ebaedbd3470f1fcebad3c11fcfe37bf
</blockquote></details>

Reverted changes.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/kubeflow/kubeflow/175)*
<!-- Sent from Reviewable.io -->

ankushagarwal commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 6 files at r2, 3 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/kubeflow/kubeflow/175#-L5vA25B-YqduVlm7k95:-L5z4WFNSvAsiMng6OuW:b890mnc) (no related file):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Updated a bunch of places to double quotes. Its not clear to me from the linked issue why double might be better.

If we're serious about wanting to enforce consistency we should look into using a lint check. 

`jsonnet fmt` can be used to autoformat and provide info about some formatting errors; don't think it handles quotes though.


</blockquote></details>

The issue suggests that support for single quotes to jsonnet was added as an afterthought - so I inferred using double quotes is better

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/175#-:-L5z5h-w2pQ9V4QP5m_n:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ankushagarwal commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/kubeflow/kubeflow/175#-L5vA25B-YqduVlm7k95:-L5zd8f7dM5JxWZQ4XVL:bw6yz2l) (no related file):*
<details><summary><i>Previously, ankushagarwal (Ankush Agarwal) wrote…</i></summary><blockquote>

The issue suggests that support for single quotes to jsonnet was added as an afterthought - so I inferred using double quotes is better
</blockquote></details>

`jsonnet fmt` seems to support this via the `--string-style` flag. The default action is to leave strings as is, but we can optionally enforce single or double quotes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/175)*
<!-- Sent from Reviewable.io -->

erikerlandson commented:
This is tangent to the issue of spartakus: It looks like a fair number of the code changes in this PR are non-functional and non-related to adding spartakus. Things like formatting and whitespace changes.  Keeping PRs as minimal as possible is a big help to reviewers.

jlewi commented:
@erikerlandson Your point is well taken. I refactored the configs a bit to make doing the spartakus changes a bit easier. If you or other reviewers would like I can split the refactoring into a separate PR and commit that first to make the spartakus changes easier to review.

jlewi commented:
I moved the refactoring into #284. I'll get that approved and submit it first.

jlewi commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/kubeflow/kubeflow/175#-L5vA25B-YqduVlm7k95:-L624NkmI2voPE0OY3ds:byu3z3t) (no related file):*
<details><summary><i>Previously, ankushagarwal (Ankush Agarwal) wrote…</i></summary><blockquote>

`jsonnet fmt` seems to support this via the `--string-style` flag. The default action is to leave strings as is, but we can optionally enforce single or double quotes.
</blockquote></details>

Cool. Do you want to send a PR to update the script to do autoformatting?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/175)*
<!-- Sent from Reviewable.io -->

jlewi commented:
/assign @aronchick

aronchick commented:
LGTM

aronchick commented:
/lgtm

jlewi commented:
@aronchick Per our discussion I made this opt in.

jlewi commented:
/approve

elmiko commented on user_guide.md:
````diff
@@ -77,6 +83,36 @@ At any time you can inspect the kubernetes objects definitions for a particular
+kubectl delete -n ${NAMESPACE} deploy spartakus-volunteer
+```
+
+**Reporting usage data is one of the most signifcant contributions you can make to Kubeflow; so please consider leaving it on.** This data
````
maybe instead of "consider leaving it on" say "consider turning it on", as this changed to opt-in

elmiko commented on user_guide.md:
````diff
@@ -77,6 +83,36 @@ At any time you can inspect the kubernetes objects definitions for a particular
+```
+
+**Reporting usage data is one of the most signifcant contributions you can make to Kubeflow; so please consider leaving it on.** This data
+allows us to improve the product and helps the many companies working on Kubeflow justify continued investement. 
````
seems like "project" might be a better word instead of "product" here

elmiko commented on user_guide.md:
```diff
@@ -77,6 +83,36 @@ At any time you can inspect the kubernetes objects definitions for a particular
 
+### Usage Reporting
+
+By default, Kubeflow will report **anonymous** usage data using [spartakus](https://github.com/kubernetes-incubator/spartakus), Kubernetes' reporting tool. Spartakus **does not report any personal information**. See [here](https://github.com/kubernetes-incubator/spartakus) for more detail.
```
maybe this should read "When enabled, Kubeflow will report ..." ?

elmiko commented on user_guide.md:
```diff
@@ -77,6 +83,36 @@ At any time you can inspect the kubernetes objects definitions for a particular
+### Usage Reporting
+
+By default, Kubeflow will report **anonymous** usage data using [spartakus](https://github.com/kubernetes-incubator/spartakus), Kubernetes' reporting tool. Spartakus **does not report any personal information**. See [here](https://github.com/kubernetes-incubator/spartakus) for more detail.
+This is entirely voluntary and you can opt out by doing the following
```
considering the opt-in/out discussion, i think we should just say "and you can disable reporting by doing the following"

jlewi commented:
@elmiko  Addressed your comments please take a look.

elmiko commented:
lgtm, thanks @jlewi !

jlewi commented:
/approve

jlewi commented:
Review status: 2 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[user_guide.md, line 88 at r7](https://reviewable.io:443/reviews/kubeflow/kubeflow/175#-L5U53-LErgYChauZ6An-r7-88:-L6Yle-jM0FFOllrfxKJ:b-896fix) ([raw file](https://github.com/kubeflow/kubeflow/blob/ec375c77f1bea94308ff7b900ee2a4abafbbddb8/user_guide.md#L88)):*
<details><summary><i>Previously, elmiko (michael mccune) wrote…</i></summary><blockquote>

maybe this should read "When enabled, Kubeflow will report ..." ?
</blockquote></details>

Done.

---

*[user_guide.md, line 89 at r7](https://reviewable.io:443/reviews/kubeflow/kubeflow/175#-L5U53-LErgYChauZ6An-r7-89:-L6Ylfojt8SOmK8TaTWF:b-896fix) ([raw file](https://github.com/kubeflow/kubeflow/blob/ec375c77f1bea94308ff7b900ee2a4abafbbddb8/user_guide.md#L89)):*
<details><summary><i>Previously, elmiko (michael mccune) wrote…</i></summary><blockquote>

considering the opt-in/out discussion, i think we should just say "and you can disable reporting by doing the following"
</blockquote></details>

Done.

---

*[user_guide.md, line 107 at r7](https://reviewable.io:443/reviews/kubeflow/kubeflow/175#-L5U53-LErgYChauZ6An-r7-107:-L6Ylg3_1c4oK8tSKJMH:b-896fix) ([raw file](https://github.com/kubeflow/kubeflow/blob/ec375c77f1bea94308ff7b900ee2a4abafbbddb8/user_guide.md#L107)):*
<details><summary><i>Previously, elmiko (michael mccune) wrote…</i></summary><blockquote>

maybe instead of "consider leaving it on" say "consider turning it on", as this changed to opt-in
</blockquote></details>

Done.

---

*[user_guide.md, line 108 at r7](https://reviewable.io:443/reviews/kubeflow/kubeflow/175#-L5U53-LErgYChauZ6An-r7-108:-L6YlgLYzBgobgyxNccM:b-896fix) ([raw file](https://github.com/kubeflow/kubeflow/blob/ec375c77f1bea94308ff7b900ee2a4abafbbddb8/user_guide.md#L108)):*
<details><summary><i>Previously, elmiko (michael mccune) wrote…</i></summary><blockquote>

seems like "project" might be a better word instead of "product" here
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/175)*
<!-- Sent from Reviewable.io -->

BenTheElder commented:
@jlewi this PR has a non-green status `code-review/reviewable` so tide will not merge it.

jlewi commented:
@BenTheElder Thanks.
